### PR TITLE
kubeadm: update the upgrade documentation for 1.20

### PR DESCRIPTION
**Pull request by neolit123**
```
layouts/shortcodes/skew.html: add latestVersionAddMinor …

For site.Params.latest equals v1.20,

The following shortcode:
  skew latestVersionAddMinor N "-"

would return:
  1-(20+N)

negative N are supported.
--------------
kubeadm: upgrade the upgrade documentation for 1.20 …

Update the kubeadm upgrade documentation:
- Re-order the steps to drain / upgrade kubelet/kubectl / uncordon
- Remove the detailed output from apply and plan.
- Use {{ skew ... }} markers so that we can avoid more verbose changes
to the document.
- Apply minor / general cleanup (whitespace, missing ':')
```

these changes as is, would result in the kubeadm team no longer having to update the documentation on every release, unless needed.

fixes https://github.com/kubernetes/kubeadm/issues/2324

/sig cluster-lifecycle

### CODE REVIEW

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:7]: Remember to
```suggestion
min-kubernetes-server-version: 1.20
```
:wink:

**Comment by annajung** [comment]
Hi @neolit123  

Friendly reminder that docs ready for review deadline is coming up on Nov 23rd, right after Kubecon NA. We would appreciate it if you could provide some/all of the doc content for this feature early so that it can be reviewed by Sig Docs before Kubecon and other events happening in November.

When it's ready to be reviewed, please don't forget to change the stage from draft to ready to be reviewed. Thank you!

**Comment by neolit123** [comment]
thanks for the reminder, @annajung 
i will add the changes next week.

**Comment by neolit123** [comment]
i cannot seem to be able to see the up to date docs preview.
does anyone know why it's not triggering?

/hold

**Comment by annajung** [comment]
hi @neolit123 I see this in the build log

```
4:01:05 PM: Error: Error building site: "/opt/build/repo/content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:2:1": failed to unmarshal YAML: yaml: invalid map key: map[interface {}]interface {}{"< skew latestVersion >":interface {}(nil)}
```

Not sure exactly what line it's pointing to, but hope this helps!

**Comment by neolit123** [comment]
hm, checking if it's due to the `{{ skew ...}}` in the yaml header.

**Comment by annajung** [comment]
ohh also, the link that you shared is the wrong one, 
you should be looking at https://deploy-preview-24908--kubernetes-io-vnext-staging.netlify.app/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/ for 1.20

**Comment by neolit123** [comment]
/cc @fabriziopandini 
for tech review.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (annajung) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:44]: just a small typo here

```suggestion
- You can only upgrade from one MINOR version to the next MINOR version,
```

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md]: added `"{{< skew latestVersion >}}"` here, instead of the literal version.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:7]: Does this actually render OK? To check, add the `{{< version-check >}}` shortcode into the page.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:22]: The existing unordered list seems better for readers - a quick scan, one click, not much thinking.

Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:22]: agreed that this degrades the UX a little

unless there is a way to automate the list links, i suggest we log an issue at k/website about it as one of the goals of the PR is to remove the kubeadm maintainer commitment for sending a kubeadm upgrade PR for the docs on every release.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:7]: how is `min-kubernetes-server-version` used on this page and do we need it?
when i was checking the render yesterday i did not see indications of it.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:22]: actually, i will try to construct the links in this PR.

for the versions in skew, we could use skew + replace for that:
https://github.com/kubernetes/website/blob/8e2544f1902d61795e62db1125d0d5d4a8157664/layouts/shortcodes/skew.html#L49-L51
https://gohugo.io/functions/replace/

so maybe something like:

```
- [Upgrading kubeadm cluster from {{< skew oldestMinorVersion >}} to {{< skew prevMinorVersion >}}](https://v{{ replace {{< skew prevMinorVersion >}} "." "-" }}.docs.kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/)
```

so if the current doc is 19->20 we can also have the link for 18->19 from skew, but not the links for the older versions
i guess we can try Hugo's math for that https://gohugo.io/functions/math/

Comment (sftim) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:22]: :+1: for logging an issue and dealing with that later.

I could have a go at a shortcode to repeat a link across all supported versions, if they all have the same path.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (sftim) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:7]: `min-kubernetes-server-version` is used for `{{< version-check >}}` and is also available for use elsewhere.

I think given this PR and its intent, I would remove that line from the YAML.

Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md]: removed.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:22]: @sftim i've added a new commit that adds `skew latestVersionAddMinor <number> "separator"`.
it makes it possible to produce any version with a custom separator.

LMK if an issue should still be logged.
we should note that the general logic in "skew.html" as is, will fail once k8s moves to v2.0 some day.

**Comment by sftim** [comment]
:+1: Looks good to get tech-reviewed.

**Review by fabriziopandini — COMMENTED** [review]
Overall lgtm from my side, only few suggestions

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:13]: i agree that for a good set of prior versions, the process has normalized to the point where we can do that. unclear about future versions (e.g. kube-proxy DS vs SPod).

so i can see us doing:
> general overview ...
> 
> automated table with versions where this is supported
> vX.YY -> vX.YY+1
> ...

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:269]: there is this bellow already:

    ```shell
    sudo systemctl daemon-reload
    sudo systemctl restart kubelet
    ```

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md]: @fabriziopandini as @sftim noted it's used with the `{{< version-check >}}` functionality to indicate to users if a certain feature requires a minimum k8s version.

it's not needed in the doc for the time being.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:159]: added general note in "additional information"

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:250]: added general note in "additional information"

**Comment by neolit123** [comment]
i'm seeing a problem where the skew shortcode inside code blocks (e.g. `shell`) is terminating the code block, see here:
https://deploy-preview-24908--kubernetes-io-vnext-staging.netlify.app/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#upgrade-kubeadm

`# replace x in 1.20`

**Comment by sftim** [comment]
> the `skew` shortcode inside code blocks

It might be that it's an outer tab shortcode that is the problem, rather than the code-block-ness.

**Comment by sftim** [comment]
Maybe someone can troubleshoot `tab` but if the issue is there, time is tight to get that fixed. Maybe use a verbatim “1.20” as an interim measure.

**Comment by neolit123** [comment]
bellow this text:
> Choose a version to upgrade to, and run the appropriate command. For example:

i see the that the shortcode adds a new line:
```
# replace x with the patch version you picked for this upgrade
sudo kubeadm upgrade apply v1.20
.x
```

**Comment by sftim** [comment]
> the shortcode adds a new line

That's fixable, https://gohugo.io/templates/introduction/#whitespace explains how to mark templates as not adding whitespace.

**Comment by neolit123** [comment]
lookes like `skew latestVersion` already has `{{- -}}`:
https://github.com/kubernetes/website/blob/master/layouts/shortcodes/skew.html#L30-L32

**Comment by neolit123** [comment]
i think the problem is this:
https://github.com/gohugoio/hugo/issues/1753#issuecomment-515360256
https://github.com/kubernetes/website/pull/24908/commits/30e2ce923c86c054dc2b55361015df5a643999fb#diff-54d8f33eefc8b623f260ed41107fd8938eb02852f7c35a1303f2c71aebe359c7L52

testing.

**Comment by neolit123** [comment]
> i think the problem is this:
gohugoio/hugo#1753 (comment)
30e2ce9#diff-54d8f33eefc8b623f260ed41107fd8938eb02852f7c35a1303f2c71aebe359c7L52

that did it.

**Comment by neolit123** [comment]
after @fabriziopandini 's recommendation here:
https://github.com/kubernetes/website/pull/24908#discussion_r525301437

it felt like a note is required:
```
{{< note >}}
This guide will use version {{< skew latestVersion >}} as an example upgrade target.
Replace {{< skew latestVersion >}} with the version you wish to upgrade to in all examples.
{{</ note >}}
```

File: layouts/shortcodes/matchingtext.html
Comment (neolit123) [inline layouts/shortcodes/matchingtext.html:1]: note: lack of new lines is intentional in the shortcode files is intentional.

**Comment by neolit123** [comment]
this is ready for another pass on reviews.

File: layouts/shortcodes/matchingtextbutton.html
Comment (sftim) [inline layouts/shortcodes/matchingtextbutton.html:1]: We can shift this script to the page footer (using `hasShortcode`) but I'm happy to leave that for a later PR.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:47]: Typo: “bellow”.

**Review by sftim — CHANGES_REQUESTED** [review]
I'm concerned that this page adds a second version selection mechanism (the one at the top right of the page is there already).

Maybe those buttons could link to the relevant page instead of revising the page interactively?

**Comment by sftim** [comment]
How about adding a shortcode that renders as `v1.20.x-00` or `1.20.x-00` as needed. Would that help?

File: layouts/shortcodes/matchingtextbutton.html
Comment (neolit123) [reply layouts/shortcodes/matchingtextbutton.html:1]: i'm not sure how that is done. tried search for `hasShortcode` online.

**Comment by neolit123** [comment]
> I'm concerned that this page adds a second version selection mechanism (the one at the top right of the page is there already).
>
> Maybe those buttons could link to the relevant page instead of revising the page interactively?

this renders the new matchingtext* shortcodes redundant, which is fine.
we can just bring back the old links (not via buttons) if that is preferred by the docs team?

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:47]: :+1:  will update once we decide on:
https://github.com/kubernetes/website/pull/24908#issuecomment-730515093

**Comment by neolit123** [comment]
> How about adding a shortcode that renders as v1.20.x-00 or 1.20.x-00 as needed. Would that help?

these version formats differ between slightly between apt and yum (x-00 vs x-0).

File: layouts/shortcodes/matchingtextbutton.html
Comment (sftim) [reply layouts/shortcodes/matchingtextbutton.html:1]: Oops, it's `.HasShortcode`: https://godoc.org/github.com/gohugoio/hugo/hugolib#Page.HasShortcode

Comment (sftim) [reply layouts/shortcodes/matchingtextbutton.html:1]: https://gohugo.io/templates/shortcode-templates/#checking-for-existence

**Comment by sftim** [comment]
>  we can just bring back the old links (not via buttons) if that is preferred by the docs team?

let's do that to get this over the line for the v1.20 release - happy to revisit post release!

File: layouts/shortcodes/matchingtextbutton.html
Comment (neolit123) [reply layouts/shortcodes/matchingtextbutton.html:1]: should it look something like this?

```
{{- if .HasShortcode "matchingtextbutton" -}}
<script src="{{"/js/matchingtext.js" | urlize | relURL }}"></script>
{{- end -}}
```

also on which line to add it in?
https://github.com/kubernetes/website/blob/0625ced4668b819eb9789570d8b9f2c5f9fef4b9/layouts/partials/footer.html

**Comment by neolit123** [comment]
>> we can just bring back the old links (not via buttons) if that is preferred by the docs team?
> let's do that to get this over the line for the v1.20 release - happy to revisit post release!

ok, sounds good.

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:22]: After the v1.20 release this will become a broken link, I think.

**Review by sftim — COMMENTED** [review]
LGTM

Ready for tech review?

**Comment by neolit123** [comment]
updated to bring back the old links but with automatic versions. removed the matchtext* shortcodes.

preview:
https://deploy-preview-24908--kubernetes-io-vnext-staging.netlify.app/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade

File: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md
Comment (neolit123) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:22]: are the `v1-16` docs planned for removal with the release of 1.20?
i think what would be removed with the release of 1.20 is `v1-15`

for the 1.20 version of this page the link would be 1.20-4 = 16:
https://v1-16.docs.kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/

Comment (sftim) [reply content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md:22]: D'oh yes.

**Comment by neolit123** [comment]
> Ready for tech review?

will ask @fabriziopandini again.

**Review by fabriziopandini — COMMENTED** [review]
@neolit123 thanks for this PR, our end-of-cycle task list is now simpler!
/lgtm

**Comment by annajung** [comment]
thanks, everyone! 

/approve

**Pull request merged**
